teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,502 @@
+```
+In this way, every time when Tekton Chains tries to support a new type of artifacts, users can form in-toto provenance from this schema format, without needing to change Tekton Chains much. In general, Tekton Chains itself doesn't want to distinguish the types of artifacts from attestation generation point of view. However, this can be used for tracking the artifact metrics, e.g. how many artifacts of a particular type that gets produced.
+
+#### Optional Alternatives for Schemas
+It will be more flexible if users can define the provenance following the defined schemas, but don't necessarily have to follow the naming patterns. Users can provide a result in array type, with name consistent with the signable target, and have a list of result names that are under that category. An example can be like:
+``` yaml
+results:
+        - name: SIGNABLE-VCS
+          value: ["FOO", "BAR"]
+        - name: FOO
+          value:
+            src_type: git
+            uri: git+https://github.com/foo/bar.git
+            digest_ago: sha256
+            digest: abc
+        - name: SIGNABLE-OCI_IMAGE
+          value: ["IMAGE1"]
+        - name: IMAGE1
+           value:
+             name: gcr.io/somerepo/someimage
+             digest: abc
+             alg: sha256
+```
+And the result that contains the provenance still need to be aligned with the schema we defined in Chains. This way, users don't need to name the results.
+This is not chosen because Chains would know the differences between different types of artifacts.
+
+## Design Evaluation
+
+### Reusability
+
+- Pro:  the defined schema can be used for all targets that can be either subjects or materials in intoto provenances
+- Con: users need to follow the exact schema in order to let Chains capture the signable targets
+
+### Simplicity
+
+- Pro: the design is supporting the same sets of json schema described in TEP 75 and 76
+- Pro: the provenance can be retrieved much more easily than the current type hinting method
+- Con: the schema can potentially grow much richer and size limit can become a problem
+
+
+### Flexibility
+
+- Pro: when supporting new types of signable targets, the schema can be easily added.
+- Con: for each new type of signable targets, the schema needs to be created individually
+
+<!--
+https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility
+
+- Are there dependencies that need to be pulled in for this proposal to work? What support or maintenance would be
+required for these dependencies?
+- Are we coupling two or more Tekton projects in this proposal (e.g. coupling Pipelines to Chains)?
+- Are we coupling Tekton and other projects (e.g. Knative, Sigstore) in this proposal?
+- What is the impact of the coupling to operators e.g. maintenance & end-to-end testing?
+- Are there opinionated choices being made in this proposal? If so, are they necessary and can users extend it with
+their own choices?
+-->
+
+### User Experience
```

Comment:
Would you be able to try out and link to an example of a TaskRun that generates this kind of result output?
This will probably be a good exercise to see what's easy / hard to do as a TaskRun author. 

For example - I don't know if we have a good way of passing in result names to TaskRuns, so generating the INPUTS/OUTPUTS lists might actually be tricky to implement as an author.

Reply:
Added an example at https://github.com/ywluogg/oci-image-structured-results